sfdx-source/apex-common/main/classes/fflib_SObjects.cls

```diff
@@ -0,0 +1,131 @@
+ *      this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ *      this list of conditions and the following disclaimer in the documentation
+ *      and/or other materials provided with the distribution.
+ * - Neither the name of the FinancialForce.com, inc nor the names of its contributors
+ *      may be used to endorse or promote products derived from this software without
+ *      specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ *  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ *  THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ *  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ *  OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**/
+public virtual class fflib_SObjects
+		extends fflib_Objects
+		implements fflib_ISObjects
+{
+
+	public Schema.DescribeSObjectResult SObjectDescribe {get; private set;}
+
+	/**
+	 * Class constructor
+	 */
+	public fflib_SObjects(List<Object> objects)
+	{
+		// todo - assert for SObjectType
+		this((List<SObject>) objects);
+	}
+
+	public fflib_SObjects(List<SObject> records)
+	{
+		super(records);
+	}
+
+	public fflib_SObjects(List<SObject> records, Schema.SObjectType sObjectType)
+	{
+		super(records);
+		SObjectDescribe = sObjectType.getDescribe();
+	}
+
+	public virtual List<SObject> getRecords()
+	{
+		return (List<SObject>) getObjects();
+	}
+
+	public virtual override Object getType()
+	{
+		return getSObjectType();
+	}
+
+	public virtual SObjectType getSObjectType()
+	{
+		return SObjectDescribe.getSObjectType();
+	}
+
+	protected virtual Set<Object> getFieldValues(Schema.SObjectField sObjectField)
+	{
+		Set<Object> result = new Set<Object>();
+		for (SObject record : getRecords())
+		{
+			result.add(record.get(sObjectField));
+		}
+		return result;
+	}
+
+	public virtual Schema.SObjectType getObjectType()
+	{
+		return SObjectDescribe.getSObjectType();
+	}
+
+	protected virtual fflib_SObjects selectByFieldValue(Schema.SObjectField sObjectField, Object value)
+	{
+		return selectByFieldValues(
+				sObjectField,
+				new Set<Object>{ value }
+		);
+	}
+
+	protected virtual fflib_SObjects selectByFieldValues(Schema.SObjectField sObjectField, Set<Object> values)
+	{
+		List<SObject> result = new List<SObject>();
+		for (SObject record : getRecords())
+		{
+			if (values.contains(record.get(sObjectField)))
```

Comment:
Please add a safe navigation operator here between the "values" and the "contains method"